File: src/main/java/ar/com/flow/money/Money.java
package ar.com.flow.money;

import lombok.EqualsAndHashCode;
import lombok.Getter;

@Getter
@EqualsAndHashCode
public class Money {
    private final String currency;
    private final int amount;

    public Money(String currency, int amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount must be non-negative: " + amount);
        }

        this.currency = currency;
        this.amount = amount;
    }

    public boolean isGreaterThanOrEqual(Money other) {
        return amount >= other.getAmount();
    }

    public boolean isLessThan(Money other) {
        return amount < other.getAmount();
    }

    public Money plus(Money amountToAdd) {
        return new Money(currency, amount + amountToAdd.getAmount());
    }

    public Money minus(Money amountToSubtract) {
        if (this.isLessThan(amountToSubtract)) {
            throw new InsufficientFundsException();
        }

        return new Money(currency, amount - amountToSubtract.getAmount());
    }
}

File: src/main/kotlin/ar/com/flow/money/Money.kt
package ar.com.flow.money

import ar.com.flow.bankaccount.domain.Currency

data class Money(val currency: Currency, val amount: Int) {
    init {
        require(amount >= 0) { "Amount must be non-negative: $amount" }
    }

    fun isGreaterThanOrEqual(other: Money): Boolean {
        return amount >= other.amount
    }

    fun isLessThan(other: Money): Boolean {
        return amount < other.amount
    }

    operator fun plus(amountToAdd: Money): Money {
        return Money(currency, amount + amountToAdd.amount)
    }

    operator fun minus(amountToSubtract: Money): Money {
        if (isLessThan(amountToSubtract)) {
            throw InsufficientFundsException()
        }
        return Money(currency, amount - amountToSubtract.amount)
    }

    companion object {
        fun zero(currency: Currency): Money {
            return Money(currency, 0)
        }
    }
}

object Dollars {
    fun amount(amount: Int): Money {
        return Money(Currency.USD, amount)
    }
}

object Pesos {
    fun amount(amount: Int): Money {
        return Money(Currency.ARS, amount)
    }
}
